### PR TITLE
Combine ammo and body parts

### PR DESCRIPTION
**Pull request by silverflyone**
PR for https://github.com/DFHack/dfhack/issues/3215.

Combine ammo, for bolts and arrows. categories: type + material + quality (+ created by). Created by is used for masterwork items.

Items which are marked for melting, is forbidden or hidden are excluded from being combined. This is in addition to the existing restrictions.

Edit: Added https://github.com/DFHack/dfhack/issues/3220.

### CODE REVIEW

**Comment by silverflyone** [comment]
@myk002 the combine ammo is ready for review. This is a small change over #676.

**Comment by silverflyone** [comment]
pre-commit.ci autofix

**Comment by silverflyone** [comment]
pre-commit.ci autofix

**Comment by silverflyone** [comment]
@myk002 Added body parts combine for review; includes butchered items only.

File: combine.lua [hunk 1]
```diff
@@ -14,10 +14,16 @@ local opts, args = {
 
 -- default max stack size of 30
 local MAX_ITEM_STACK=30
+local MAX_AMMO_STACK=25
 local MAX_CONT_ITEMS=500
+local MAX_MAT_AMT=500
 
 -- list of types that use race and caste
 local typesThatUseCreatures={REMAINS=true,FISH=true,FISH_RAW=true,VERMIN=true,PET=true,EGG=true,CORPSE=true,CORPSEPIECE=true}
+local typesThatUseMaterial={CORPSEPIECE=true}
```
Comment (myk002) [inline combine.lua:23]: for this and the line above, `utils.invert()` provides a cleaner syntax:
```lua
local typesThatUseCreatures = utils.invert{'REMAINS', 'FISH', 'FISH_RAW', 'VERMIN', 'PET', 'EGG', 'CORPSE', 'CORPSEPIECE'}
```

File: docs/combine.rst [hunk 2]
```diff
@@ -53,11 +55,33 @@ Options
...
+on fire, encased, owned by a trader/hostile/dwarf or is in a spider web.
+4. An item is a part and not butchered.
+
+The following categories are used for combining:
+1. Item is a part and has a race: category=type + race
+2. Item has a race/caste: category=type + race + caste
+3. Item is ammo, created by for masterwork. category=type + material + quality (+ created by)
+4. Or: category= type + material
+
+A default stack size of 30 applies to a category, unless a larger stack exists.
```
Comment (myk002) [inline docs/combine.rst:87]:
```suggestion
Each category has a default stack size of 30 unless a larger stack already exists in your fort.
In that case the largest existing stack size is used.
```

File: docs/combine.rst [hunk 3]
```diff
@@ -53,11 +55,33 @@ Options
...
+2. An item is sand or plaster.
+3. An item is rotten, forbidden/hidden, marked for dumping/melting,
+on fire, encased, owned by a trader/hostile/dwarf or is in a spider web.
+4. An item is a part and not butchered.
+
+The following categories are used for combining:
+1. Item is a part and has a race: category=type + race
+2. Item has a race/caste: category=type + race + caste
+3. Item is ammo, created by for masterwork. category=type + material + quality (+ created by)
+4. Or: category= type + material
```
Comment (myk002) [inline docs/combine.rst:85]:
```suggestion
The following categories are defined:
1. Corpse pieces, grouped by piece type and race
2. Items that have an associated race/caste, grouped by item type,  race, and caste
3. Ammo, grouped by ammo type, material, and quality. If the ammo is a masterwork, it is
    also grouped by who created it.
4. Anything else, grouped by item type and material
```

Comment (myk002) [reply docs/combine.rst:85]: Did I successfully represent your intended meaning?

File: docs/combine.rst
[the remark this answers is not in the document]
Comment (silverflyone) [reply docs/combine.rst:1]: Sorry, I missed that one. Done.

File: docs/combine.rst (hunk 2 above)
Comment (silverflyone) [reply docs/combine.rst:87]: Done.

File: docs/combine.rst (hunk 3 above)
Comment (silverflyone) [reply docs/combine.rst:85]: Looks good. Done.

**Review by myk002 — APPROVED** [review]

**Comment by myk002** [comment]
pre-commit.ci autofix

**Pull request merged**